File: Shared/CPP/HCMSessionManager.cpp
//
//  HCMSessionManager.m
//  CPP
//
//  Copyright (c) 2015 Harry Cheung. All rights reserved.
//

#include "HCMSessionManager.h"
#include <iostream>

HCMSessionManager *HCMSessionManager::instance = new HCMSessionManager();

HCMSessionManager::HCMSessionManager() {
  session = NULL;
}

void HCMSessionManager::start(HCMTrack *track, double startTime) {
  if (session == NULL) {
    this->track = track;
    session = new HCMSession(track, startTime);
    currentLap = new HCMLap(session, track, startTime, 0);
    session->laps.push_back(currentLap);
    splits = new double[track->gates->size()];
    splitGaps = new double[track->gates->size()];
    splitStartTime = startTime;
    splitNumber = 0;
    currentSplit = 0;
    lapNumber = 0;
    gap = 0;
    bestIndex = 0;
    nextGate = track->start;
    bestLap = NULL;
  }
}

void HCMSessionManager::end() {
  if (session != NULL) {
    delete session;
    delete [] splits;
    delete [] splitGaps;
    session = NULL;
  }
}

void HCMSessionManager::gps(double latitude, double longitude, double speed, double bearing,
                            double horizontalAccuracy, double verticalAccuracy, double timestamp) {
  HCMPoint point = HCMPoint(latitude, longitude, false, speed, bearing,
                            horizontalAccuracy, verticalAccuracy, timestamp);
  if (currentLap->points.size() != 0) {
    HCMPoint cross = HCMPoint();
    if (nextGate->crossed(lastPoint, point, cross)) {
      currentLap->add(cross);
      currentLap->splits[currentSplit] = cross.splitTime;
      switch (nextGate->gateType) {
        case HCMGate::START_FINISH:
        case HCMGate::FINISH:
          if (currentLap->points[0].generated) {
            currentLap->valid = true;
            if (bestLap == NULL || currentLap->duration < bestLap->duration) {
              bestLap = currentLap;
            }
          }
        case HCMGate::START:
          lapNumber++;
          currentLap =  new HCMLap(session, track, cross.timestamp, lapNumber);
          lastPoint = HCMPoint(cross.latitudeDegrees(), cross.longitudeDegrees(),
                               false, cross.speed, cross.bearing,
                               cross.hAccuracy, cross.vAccuracy, cross.timestamp);
          lastPoint.lapDistance = 0;
          lastPoint.lapTime = 0;
          lastPoint.generated = true;
          currentLap->add(lastPoint);
          session->laps.push_back(currentLap);
          gap = 0;
          for (int i = 0; i < track->gates->size(); i++) {
            splitGaps[i] = 0.0;
          }
          bestIndex = 0;
          currentSplit = 0;
          break;
        case HCMGate::SPLIT:
          if (bestLap != NULL) {
            splitGaps[currentSplit] = currentLap->splits[currentSplit] - bestLap->splits[currentSplit];
          }
          currentSplit++;
      }
      splitStartTime = cross.timestamp;
      nextGate = track->gates->operator[](currentSplit);
    }
    if (bestLap != NULL && bestIndex < bestLap->points.size()) {
      while (bestIndex < bestLap->points.size()) {
        HCMPoint refPoint = bestLap->points[bestIndex];
        if (refPoint.lapDistance > currentLap->distance) {
          HCMPoint lastRefPoint = bestLap->points[bestIndex - 1];
          double distanceToLastRefPoint = currentLap->distance - lastRefPoint.lapDistance;
          if (distanceToLastRefPoint > 0) {
            double sinceLastRefPoint = distanceToLastRefPoint / point.speed;
            gap = point.lapTime - sinceLastRefPoint - lastRefPoint.lapTime;
            splitGaps[currentSplit] = point.splitTime - sinceLastRefPoint - lastRefPoint.splitTime;
          }
          break;
        }
        bestIndex++;
      }
    }
    point.lapDistance = lastPoint.lapDistance + lastPoint.distance(point);
    point.setLapTime(currentLap->startTime, splitStartTime);
  }
  currentLap->add(point);
  lastPoint = point;
}

File: Shared/ObjC/HCMSessionManager.m
//
//  HCMSessionManager.m
//  ObjC
//
//  Created by harry on 2/27/15.
//  Copyright (c) 2015 Harry Cheung. All rights reserved.
//

#import "HCMSessionManager.h"

@implementation HCMSessionManager

+ (id)instance {
  static HCMSessionManager *sharedInstance = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    sharedInstance = [[self alloc] init];
  });
  return sharedInstance;
}

- (id)init {
  return [super init];
}


- (void)startSessionWithTrack:(HCMTrack *)track {
  if (_session == nil) {
    _track = track;
    _session = [[HCMSession alloc] initWithTrack:track];
    _currentLap = [[HCMLap alloc] initWithSession:_session
                                            track:track
                                        startTime:_session.startTime
                                        lapNumber:0];
    [_session.laps addObject:_currentLap];
    _splits = malloc([track numSplits] * sizeof(double));
    _splitGaps = malloc([track numSplits] * sizeof(double));
    _splitStartTime = _session.startTime;
    _splitNumber = 0;
    _currentSplit = 0;
    _lapNumber = 0;
    _gap = 0;
    _bestIndex = 0;
    _nextGate = track.start;
    _bestLap = nil;
    _lastPoint = nil;
  }
}

- (void)endSession {
  if (_session != nil) {
    [_session release];
    free(_splits);
    free(_splitGaps);
    [_currentLap release];
    _session = nil;
    _currentLap = nil;
    _bestLap = nil;
  }
}

- (void)gpsWithLatitude:(double)latitude
              longitude:(double)longitude
                  speed:(double)speed
                bearing:(double)bearing
     horizontalAccuracy:(double)horizontalAccuracy
       verticalAccuracy:(double)verticalAccuracy
              timestamp:(double)timestamp {
  HCMPoint *point = [[[HCMPoint alloc] initWithLatitude:latitude
                                             longitude:longitude
                                                 speed:speed
                                               bearing:bearing
                                    horizontalAccuracy:horizontalAccuracy
                                      verticalAccuracy:verticalAccuracy
                                             timestamp:timestamp] autorelease];
  if (_lastPoint != nil) {
    HCMPoint *cross = [_nextGate crossedWithStart:_lastPoint destination:point];
    if (cross != nil) {
      [_currentLap addPoint:cross];
      _currentLap.splits[_currentSplit] = cross->splitTime;
      switch (_nextGate->type) {
        case START_FINISH:
        case FINISH:
          if (((HCMPoint *)_currentLap.points[0])->generated) {
            _currentLap.valid = true;
            if (_bestLap == nil || _currentLap.duration < _bestLap.duration) {
              _bestLap = _currentLap;
            }
          }
        case START:
          _lapNumber++;
          [_currentLap release];
          _currentLap = [[HCMLap alloc] initWithSession:_session
                                                  track:_track
                                              startTime:cross->timestamp
                                              lapNumber:_lapNumber];
          _lastPoint = [[[HCMPoint alloc] initWithLatitude:[cross latitudeDegrees]
                                                longitude:[cross longitudeDegrees]
                                                    speed:cross->speed
                                                  bearing:cross->bearing
                                       horizontalAccuracy:cross->hAccuracy
                                         verticalAccuracy:cross->vAccuracy
                                                timestamp:cross->timestamp] autorelease];
          _lastPoint->lapDistance = 0;
          _lastPoint->lapTime = 0;
          _lastPoint->generated = true;
          [_currentLap addPoint:_lastPoint];
          [_session.laps addObject:_currentLap];
          _gap = 0;
          for (int i = 0; i < [_track numSplits]; i++) {
            _splitGaps[i] = 0.0;
          }
          _bestIndex = 0;
          _currentSplit = 0;
          break;
        case SPLIT:
          if (_bestLap != nil) {
            _splitGaps[_currentSplit] = _currentLap.splits[_currentSplit] - _bestLap.splits[_currentSplit];
          }
          _currentSplit++;
      }
      _splitStartTime = cross->timestamp;
      _nextGate = _track.gates[_currentSplit];
    }
    if (_bestLap != nil && _bestIndex < _bestLap.points.count) {
      while (_bestIndex < _bestLap.points.count) {
        HCMPoint *refPoint = _bestLap.points[_bestIndex];
        if (refPoint->lapDistance > _currentLap.distance) {
          HCMPoint *lastRefPoint = _bestLap.points[_bestIndex - 1];
          double distanceToLastRefPoint = _currentLap.distance - lastRefPoint->lapDistance;
          if (distanceToLastRefPoint > 0) {
            double sinceLastRefPoint = distanceToLastRefPoint / point->speed;
            _gap = point->lapTime - sinceLastRefPoint - lastRefPoint->lapTime;
            _splitGaps[_currentSplit] = point->splitTime - sinceLastRefPoint - lastRefPoint->splitTime;
          }
          break;
        }
        _bestIndex++;
      }
    }
    point->lapDistance = _lastPoint->lapDistance + [_lastPoint distanceToPoint:point];
    [point setLapTimeWithStartTime:_currentLap.startTime splitStartTime:_splitStartTime];
  }
  [_currentLap addPoint:point];
  _lastPoint = point;
}

@end
